### PR TITLE
Add string-refs support with shallowWrapper

### PR DESCRIPTION
**Pull request by yormi**
In response to https://github.com/airbnb/enzyme/issues/361 and my own need :)

I find it extremely useful to have a unique, local-scoped, way to identify my React Element.

Works as expected `shallowWrapper.ref('stringRef') => ShallowWrapper` or throws if more than one element has the wanted ref.

### CODE REVIEW

File: src/ShallowTraversal.js [hunk 1]
```diff
@@ -125,7 +124,7 @@ export function buildPredicate(selector) {
         return node => nodeMatchesObjectProps(node, selector);
       }
       throw new TypeError(
-        'Enzyme::Selector does not support an array, null, or empty object as a selector',
+        'Enzyme::Selector does not support an array, null, or empty object as a selector'
```
Comment (ljharb) [inline src/ShallowTraversal.js]: this trailing function comma is required, why did you remove it?

File: src/ShallowTraversal.js [hunk 2]
```diff
@@ -1,7 +1,6 @@
 import React from 'react';
 import isEmpty from 'lodash/isEmpty';
 import isSubset from 'is-subset';
-import functionName from 'function.prototype.name';
```
Comment (ljharb) [inline src/ShallowTraversal.js]: i think perhaps you included some unintentional changes - have you rebased on the latest master?

File: src/ShallowTraversal.js (hunk 1 above)
Comment (yormi) [inline src/ShallowTraversal.js]: I'm not sure what happened but it seems that I used some kind of auto-format that I'm not aware of... sorry for that.

Correcting that !

File: src/ShallowTraversal.js [hunk 3]
```diff
@@ -151,3 +151,23 @@ export function getTextFromNode(node) {
     .join('')
     .replace(/\s+/, ' ');
 }
+
+export const isAReactComponentNode = node => typeof node.type === 'function';
```
Comment (ljharb) [inline src/ShallowTraversal.js]: Instead of defining your own function, please use `isFunctionalComponent`, `isCustomComponentElement`, and `isDOMComponentElement` which are already present in the file.

File: src/ShallowWrapper.js [hunk 4]
```diff
@@ -454,6 +456,32 @@ class ShallowWrapper {
...
+   */
+  ref(ref) {
+    return this.single('ref', (node) => {
+      if (!isAReactComponentNode(this.unrendered)) {
+        throw new Error('The current node must be a React component.');
+      }
+
+      const result = findRefInOwnedChildren(node, ref, true);
+
+      if (result.length === 0) return false;
```
Comment (ljharb) [inline src/ShallowWrapper.js]: shouldn't this return `null` when there's no result?

File: src/ShallowWrapper.js [hunk 5]
```diff
@@ -454,6 +456,32 @@ class ShallowWrapper {
...
+      if (result.length === 1) return new ShallowWrapper(result[0], this.root);
+      if (result.length > 1) {
+        throw new Error('There is more than one component own by the tree root with the same ref string');
+      }
+
+      return false;
```
Comment (ljharb) [inline src/ShallowWrapper.js]: when is this line hit?

**Comment by yormi** [comment]
I still wonder why they've been deprecated for callbacks. I don't remember reading about the reason behind it.

I personally use the string-refs to access my react-element in integration-test. I use them to pilot my tests rather than testing the ref.

I haven't found another way that is as elegant(local-scoped, unique) yet... I'm open to other solutions though.

File: src/ShallowWrapper.js (hunk 5 above)
Comment (yormi) [inline src/ShallowWrapper.js]: In case of many "owned" element with the same ref. Shouldn't happen but if an error occurs, I think it'll be better to know it instead of returning the first occurrence.

Comment (ljharb) [inline src/ShallowWrapper.js]: The previous lines though cover all possibilities for `result.length

File: src/ShallowWrapper.js (hunk 4 above)
Comment (yormi) [inline src/ShallowWrapper.js]: You are right !

File: src/ShallowWrapper.js (hunk 5 above)
Comment (yormi) [inline src/ShallowWrapper.js]: Oupsi... you meant the `return false`. I thought the thrown error. You are right again !

File: src/ShallowTraversal.js [hunk 6]
```diff
@@ -151,3 +154,21 @@ export function getTextFromNode(node) {
     .join('')
     .replace(/\s+/, ' ');
 }
+
+export const findRefInOwnedChildren = (node, ref, isRootNode) => {
```
Comment (ljharb) [inline src/ShallowTraversal.js]: this being exported, having `isRootNode` be provideable by the consumer seems brittle. Could we split this up into separate functions, so that the recursion implementation details are hidden?

**Comment by Aweary** [comment]
> It's worth discussing that string refs are deprecated in React (in favor of callback refs)

They're not technically deprecated just yet, just considered legacy. The point still stands though, it's not a pattern that's recommended anymore.

Good work on the PR @yormi, but I would have to say that I'm not in favor of this. Adding`.ref()` to the API solely to support a legacy implementation seems confusing to me.

**Comment by yormi** [comment]
First of all, thank you :)

Secondly, I understand your concerns @Aweary. I would prefer avoiding legacy/deprecated features but I haven't found other ways as clear and simple to do so.

I would be very interested in your suggestions on a (hopefully) local-scoped and unique way of accessing an element to pilot tests ?

**Comment by Aweary** [comment]
> I would be very interested in your suggestions on a (hopefully) local-scoped and unique way of accessing an element to pilot tests ?

If you use ref callbacks you should be able to access them on the instance

```js
class Foo extends React.Component {
  render() {
    return <div ref={n => this.node = n}/>
  }
}

const wrapper = mount(<Foo />);
const div = wrapper.instance().node; // should be the HTMLDivElement instance
```

I don't think shallow should support refs, let alone string refs.

**Comment by blainekasten** [comment]
Just to ignite a flame, I would actually advocate for support for strong refs right now. Though they are legacy, they are still a supported feature. String refs are used wildly in the wild, so without the feature we're making every consumer write brittle and confusing tests to get access to their refs. Or they just don't write tests because it's too difficult. Both are problems this library should be trying to help with.

We can remove the feature as react removes it. Which would make more sense to do then as consumers updating to newer versions of react would HAVE to change their refs.

**Comment by ljharb** [comment]
@blainekasten I don't mind supporting string refs - but i *do* mind having a nicer API for string refs than for callback refs. We should be providing API that encourages people to do the right thing, even if we also provide API that enables people to test a legacy pattern.

**Comment by blainekasten** [comment]
@ljharb that totally makes sense. Do we even have a designated API to work with callback refs?

File: src/ShallowTraversal.js (hunk 6 above)
Comment (yormi) [inline src/ShallowTraversal.js]: Yeah, good idea. I'm on it. Thanks for the feedback btw. I haven't had the chance to have my code reviewed much.

**Comment by yormi** [comment]
@Aweary I agree that works (or should work... I think I saw an issue about it !?) with `mount`. I think it should be possible to use refs with `shallow` as well. If we find a way to use callback refs with `shallow`, I'd be happy to change my string refs.

I think it'll be possible:
By getting all the refs of the root (pretty much as I did) and executing them in the shallowWrapper constructor we could access the "refs" on the instance as with `mount`.

The problems I foresee are:
- if a callback is more complex than `c => { this.foo = c }` it might have code that needs to have the whole thing rendered... I guess it shouldn't happen though
- we need to find a way not to shallow render all the elements with a ref

What do you guys think ?

**Comment by Aweary** [comment]
>if a callback is more complex than c => { this.foo = c } it might have code that needs to have the whole thing rendered... I guess it shouldn't happen though`

Correct, ref callbacks shouldn't do anything to affect the render output, and it should almost always just be attaching the node to the instance.

I'm OK with adding support for refs with `shallow`, but we need to support string refs and callback refs if we do. Supporting only a legacy implementation isn't good.

> Couldn't we just let the same API resolve for both string and callbacks? I'm not sure if it's safe to assume that refs always get stored on the component instance, but we could do something like this:

The shallow renderer doesn't actually process refs, so unfortunately we can't.
>Shallow testing currently has some limitations, namely not supporting refs.


[ - source](https://facebook.github.io/react/docs/test-utils.html#shallow-rendering)

**Comment by yormi** [comment]
Ok so I went on with the idea but since I didn't find a way to make the callback refs work within `enzyme`, I dive into `react` code: https://github.com/facebook/react/pull/8311

**Comment by yormi** [comment]
Ok since the react team seems to have a lot to do with Fiber and everything, maybe it'll be simpler to have the feature implemented here. I might have found a way to make it work with callback refs as well.

The problem I'm facing is that callback refs are bound to the instance and it seems that the instance is undefined when calling it from enzyme(i.e. not directly during rendering in react ShallowRenderer).

We can still copy the render method so when the callback ref are rendered, they have a different `this`. This is possible only if the render method is not fat arrow function. I don't quite understand why but it works when it's bound using `.bind` but not when using "fat-arrow-binding".

@Aweary do you think it's fair to assume that the render method is not or shouldn't be bound using a fat-arrow function ? Too big assumption ? I hate to have limitation like that but on the other hand, I don't see why we would need an arrow function there.

**Comment by Aweary** [comment]
@yormi I went ahead and reviewed your PR on the React repo. If this is implemented in React it shouldn't require any changes to the core reconciler modules. It might be better to implement this in enzyme, if you've found a valid approach that supports callback refs.

**Comment by yormi** [comment]
Ok my laptop HDD just drown in the snowstorm here... I just replaced it. I'll try to implement it in enzyme in the next 2 days. Thanks for the feedback :)

**Comment by yormi** [comment]
Ok sorry for the delay ! Semester's over now so here's my new proposal.

Refs are accessible directly on the instance so it's easily accessible to just check the props. It's also easy to build upon if you want to add to to the API to dive in the tree... even though that's not the point of shallow rendering, I think.

**Comment by yormi** [comment]
Actually, I found a "bug"... The refs are actualized at every render, thanks to `componentDidUpdate`.

This hook is called when we update a component state through a button handler for instance but not when using `wrapper.setProps` or anything using the `ShallowWrapper.prototype.rerender`. This cause the props of the refs to not be in sync. If `lifecycleExperimental` is turned on, the `componentDidUpdate` fires and everything is alright.

So what is the purpose of having `lifecycleExperimental` flag turned off ?

Do you see a better way to listen on the component render ?

**Comment by ljharb** [comment]
Now that createRef/forwardRef is the preferred way forward for refs - even though string and callback refs are unlikely to go away - I don't think this is a feature worth making first-class in shallow rendering.

What would the ref even return for the cases where in `mount` it returns a DOM element?

I'm going to close this; but we can reopen if needed.
